### ISSUE TITLE
How to plot the average gradient?

### ISSUE THREAD

**Issue by capetocape** [Ori226/p300_lstm#1]
Hello, Ori.

I've read your paper. I want to know how to plot the average gradient? Just like FIGURE 4.1 in your paper.

Thank you !!

**Comment by Ori226** [comment]
You mean how did I implement it? Or how I did it mathematically?

**Comment by capetocape** [comment]
how to implement?
I just tried it. It doesn't work very well

**Comment by Ori226** [comment]
@capetocape , How did you try it? Anyway, I'll try to build a concrete example and uploaded it to this repository(it may take several days). Meanwhile, you can try to post your code here.

**Comment by capetocape** [comment]
```
import matplotlib.pyplot as plt
from keras import backend as K
import tensorflow as tf
def plot_gradients(model, x_train, eeg_sample_shape):
    outputTensor = model.output[0]
    variableTensors = model.input
    gradients = K.gradients(outputTensor, variableTensors)
    #
    trainingExample = x_train.reshape(1, eeg_sample_shape[0], eeg_sample_shape[1])
    sess = tf.InteractiveSession()
    sess.run(tf.initialize_all_variables())
    evaluated_gradients = sess.run(gradients, feed_dict={model.input: trainingExample})
    #
    evaluated_gradients = evaluated_gradients[0].reshape(eeg_sample_shape[0], eeg_sample_shape[1])
    plt.imshow(evaluated_gradients.T, interpolation='nearest', cmap='jet')
    plt.show()
```

When I use this code to draw the gradient, the experimental results are as follows:

![gradient_of_cnn_and_lstm](https://user-images.githubusercontent.com/23657885/50470860-4588c080-09ed-11e9-961f-f72940b4f44a.png)

(The data I used is the BCI Competition Dataset)

**Comment by Ori226** [comment]
The plot does look pretty bad. Please send me a link to the dataset. Also, doesn't the network had success in task? From what I see, it seems like the data is not channel-wise normalized.

**Comment by capetocape** [comment]
http://www.bbci.de/competition/iii/
(Data Set II)

Both models works well with this data set. (I'm using a feature-wise normalization)

**Comment by capetocape** [comment]
Is there anything wrong with my code?

**Comment by Ori226** [comment]
I'm still working on preparing the code. It seems that you don't average the gradient over all the "target" several samples. It looks you do it only on a single example.

**Comment by capetocape** [comment]
When I use multiple  “target” samples, the result is the same. As you can see from my second experiment results, the gradients along each time step are almost the same. So I think maybe there's something wrong with my code.
Could you please post the code of gradient calculation briefly first? :)

**Comment by Ori226** [comment]
I'm working on your code and I think I've spotted at least one bug:
The following two lines:
```
sess = tf.InteractiveSession()
sess.run(tf.initialize_all_variables())

```
Actually initialized the model's weight.

**Comment by Ori226** [comment]
I created the next figure with the code below:
![figure_1](https://user-images.githubusercontent.com/5808240/50605883-6446f580-0ecc-11e9-955c-3a870ede6307.png)

Please pay attention to the difference between your code. Especially the `K.get_session()`, the fact that I don't reinitialize the variable and the computation of the average gradient only over the target samples. Feel free to ask for additional help. The model was trained using `run_multi_subject_experiment.py` but with `number_of_k_fold = 2`. Good luck.

```
"""
In the following code I create a show the saliency map of the network.
"""
import keras
import matplotlib.pyplot as plt
from keras import backend as K
import tensorflow as tf
import numpy as np

from run_multi_subject_experiment import prepare_data_for_experiment


def main():

    model = keras.models.load_model(r"<PATH_TO_YOUR_MODEL>")
    add_time_domain_noise = False

    # all_subjects = ['gcd', 'fat', 'gcc', 'icr', 'icn', 'iay', 'gch', 'gcg', 'gcf', 'gcb', 'pia']

    # in this case I use only subject gcd
    all_subjects = ['gcd']

    # These are the settings the model was trained with
    cross_validation_iter = 0
    current_experiment_setting = 'Color116ms'
    downsample_params = 8
    number_of_k_fold = 2

    train_data, train_tags, test_data_with_noise, test_tags, noise_shifts = prepare_data_for_experiment(all_subjects,
                                                                                                        add_time_domain_noise=add_time_domain_noise,
                                                                                                        current_experiment_setting=current_experiment_setting,
                                                                                                        downsample_params=downsample_params,
                                                                                                        number_of_k_fold=number_of_k_fold,
                                                                                                        cross_validation_iter=cross_validation_iter)
    # Note that the shape of the data is <#trials,#attempt_per_letter,#timestamps,#channles>.
    # The following line take only the 'target' samples
    only_target_samples = train_data.reshape(-1, train_data.shape[2], train_data.shape[3])[train_tags == 1]

    average_gradent = get_average_gradient(model=model,x_train=only_target_samples)

    plt.subplot(1,2,1)
    plt.imshow(np.mean(only_target_samples, axis=0).T, interpolation='nearest', cmap='jet', aspect='auto')
    plt.title('raw signal average')
    plt.subplot(1, 2, 2)
    plt.title('gradient average')
    plt.imshow(average_gradent.T, interpolation='nearest', cmap='jet', aspect='auto')
    plt.show()


def get_average_gradient(model, x_train):
    outputTensor = model.output[0]
    variableTensors = model.input
    gradients = K.gradients(outputTensor, variableTensors)
    trainingExample = x_train
    sess = K.get_session()
    evaluated_gradients = sess.run(gradients, feed_dict={model.input: trainingExample})
    average_gradient = np.mean(np.abs(evaluated_gradients[0]), axis=0)
    return average_gradient


if __name__ == '__main__':
    main()
```
